wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointMonitorImpl.java

```diff
@@ -0,0 +1,229 @@
+    this.pluginServicesLock.readLock().lock();
+    try {
+      return this.pluginServices.isEmpty();
+    } finally {
+      this.pluginServicesLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void close() {
+    this.stop.set(true);
+
+    try {
+      // TODO: the termination timeout is taken from failover2, should it be shorter since it blocks the current thread?
+      if (!this.monitorExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+        this.monitorExecutor.shutdownNow();
+        this.rdsClient.close();
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      this.monitorExecutor.shutdownNow();
+      this.rdsClient.close();
+    }
+  }
+
+  @Override
+  public void run() {
+    try {
+      while (!this.stop.get() && !Thread.currentThread().isInterrupted()) {
+        long start = System.nanoTime();
+
+        final Filter customEndpointFilter =
+            Filter.builder().name("db-cluster-endpoint-type").values("custom").build();
+        final DescribeDbClusterEndpointsResponse endpointsResponse =
+            this.rdsClient.describeDBClusterEndpoints(
+                (builder) ->
+                    builder.dbClusterEndpointIdentifier(this.endpointIdentifier).filters(customEndpointFilter));
+        List<DBClusterEndpoint> endpoints = endpointsResponse.dbClusterEndpoints();
+        if (endpoints.size() != 1) {
+          List<String> endpointURLs = endpoints.stream().map(DBClusterEndpoint::endpoint).collect(Collectors.toList());
+          LOGGER.warning(
+              Messages.get("CustomEndpointMonitorImpl.unexpectedNumberOfEndpoints",
+                  new Object[] {
+                      this.endpointIdentifier,
+                      this.region.id(),
+                      endpoints.size(),
+                      endpointURLs
+                  }
+              ));
+
+          // TODO: should we throw an exception and then return? Or wait and then try again?
+          long elapsedTime = System.nanoTime() - start;
+          long sleepDuration = Math.min(0, this.refreshRateNano - elapsedTime);
+          TimeUnit.NANOSECONDS.sleep(sleepDuration);
+          continue;
+        }
+
+        CustomEndpointInfo endpointInfo = CustomEndpointInfo.fromDBClusterEndpoint(endpoints.get(0));
+        if (customEndpointInfoCache.get(this.endpointIdentifier).equals(endpointInfo)) {
+          long elapsedTime = System.nanoTime() - start;
+          long sleepDuration = Math.min(0, this.refreshRateNano - elapsedTime);
+          TimeUnit.NANOSECONDS.sleep(sleepDuration);
+          continue;
+        }
+
+        // The custom endpoint info has changed, so we need to update the info in the registered plugin services.
+        customEndpointInfoCache.put(this.endpointIdentifier, endpointInfo, this.cacheEntryExpirationNano);
+        this.pluginServicesLock.readLock().lock();
+        try {
+          for (PluginService service : this.pluginServices) {
```

Comment:
Each plugin service instance belongs to a wrapper connection. By keeping references to plugin services you prevent them from disposing by GC. You need to consider using weak pointers.

Reply:
Also, you call plugin services from a monitoring thread while plugin service might handle some other action so PluginServiceImpl should be multi-thread safe that we never though about. 

Is there alternative for keeping plugin services and sending them an updates?

Reply:
When connections are being closed they deregister their plugin service in releaseResources which would remove the reference and allow the GC to dispose of the connection, however I should maybe change to weak references anyways in case the connection does not cleanly close. 

On main, the PluginServiceImpl class already has a lock variable (`connectionSwitchLock`, which is used in `setCurrentConnection`), doesn't that imply the plugin service is already being used in some multi-threaded capacity? In the case of this PR, the shared data is stored in the statuses CacheMap which uses a ConcurrentHashMap, so I believe it this scenario should be thread-safe (let me know if I've missed something).

Regardless, I'm open to other suggestions if you have them. The tricky thing here is that we need a mechanism that allows each custom endpoint connection to update their custom endpoint info regularly over time. The only other way I can think of doing that is having a method that plugins or the plugin service can call to grab custom endpoint info from the cache, similar to `refresh` and `forceRefresh` in the host list provider. But plugins, and the plugin service, aren't aware of what other plugins are in use so they cannot call the custom endpoint plugin directly.

Reply:
I removed the plugin service list after further discussions with Sergiy